[Cocoa] font-variation-settings:opsz is not applied when text-rendering:optimizeLegibility is set

litherum opened the pull request:
#### 8bbfbb1f0689efdbb29611af60e552da8b16d993
<pre>
[Cocoa] font-variation-settings:opsz is not applied when text-rendering:optimizeLegibility is set
<a href="https://bugs.webkit.org/show_bug.cgi?id=254897">https://bugs.webkit.org/show_bug.cgi?id=254897</a>
rdar://106814619

Reviewed by Simon Fraser.

text-rendering:optimizeLegibility turns on kCTFontOpticalSizeAttribute:”auto” in Core Text,
and font-variation-settings:opsz turns on kCTFontVariationAttribute:opsz in Core Text. When
both are applied, font-variation-settings is supposed to win. However, in Core Text,
kCTFontOpticalSizeAttribute wins.

The solution for this is to set kCTFontOpticalSizeAttribute to a specific size, given by
font-variation-settings:opsz. This should only happen when OpticalSizingTypes::Everything
is enabled, so this patch adds an optional&lt;float&gt; to that struct. All this is guarded by
a USE() so we can remove it if/when Core text changes to match the CSS behavior.

Test: fast/text/variations/optical-sizing-optimizeLegibility.html

* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/platform/graphics/FontTaggedSettings.h:
* Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.cpp:
(WebCore::UnrealizedCoreTextFont::addAttributesForOpticalSizing):
(WebCore::UnrealizedCoreTextFont::modifyFromContext):
* Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.h:

Canonical link: <a href="https://commits.webkit.org/262831@main">https://commits.webkit.org/262831@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bcd51a2eb9ac0cb496bb9dfd2706753a065fe255

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2739 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2739 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2885 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4150 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3105 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2704 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2884 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2837 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2411 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2765 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3200 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2451 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3918 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/694 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2436 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/2298 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/2284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2821 "Built successfully") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2450 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3676 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/2613 "Built successfully and passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/2487 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/2265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/2813 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2485 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2439 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/667 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/675 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2478 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/2878 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2634 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/785 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.h:
```diff
@@ -70,16 +71,19 @@ class UnrealizedCoreTextFont {
+        // When USE(CORE_TEXT_OPTICAL_SIZING_WORKAROUND) is no longer necessary, we can migrate this back to an enum.
+        struct None { };
+        struct JustVariation { };
+        struct Everything { std::optional<float> opticalSizingValue; };
```
Could we save some memory by using a `Markable` instead of an `optional` here ? (do we actually need the 0 value of the negative values?)

litherum replied:
I think we actually do need the full range, because the value that this gets set to is supplied by CSS (parsed via `CSSPropertyParserHelpers::consumeNumberRaw()`)

mdubet commented on Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.cpp:
```diff
@@ -306,15 +306,29 @@ void UnrealizedCoreTextFont::modifyFromContext(const FontDescription& fontDescri
+        auto predicate = [opsz = fontFeatureTag("opsz")](const FontVariationSettings::Setting& setting) {
+            return setting.tag() == opsz;
+        };
+        auto iterator = std::find_if(m_variationSettings.begin(), m_variationSettings.end(), WTFMove(predicate));
```
Free form `std::begin` and `std::end` are more general than class-based one (but it's not part of the WebKit guidelines though so I guess it's not important)

litherum replied:
Done.

rreno commented on Source/WTF/wtf/PlatformUse.h:
```diff
@@ -376,3 +376,10 @@
 #endif
+
+#if (PLATFORM(MAC) && __MAC_OS_X_VERSION_MAX_ALLOWED < 150000) \
+    || (PLATFORM(IOS_FAMILY) && __IPHONE_OS_VERSION_MAX_ALLOWED < 180000) \
```
If these `*_MAX_ALLOWED` macros are undefined, they'll be treated as zero - is it possible for `PLATFORM(IOS_FAMILY)` to be defined but `__IPHONE_OS_VERSION_MAX_ALLOWED` to be undefined? Maybe in the case of a non-iPhone IOS_FAMILY OS?

Elsewhere we use `*_MAX_ALLOWED` as the left hand side of a > or >= expression which helps avoid the undefined-macro-is-zero situation from inadvertently causing the expression to evaluate to true.

litherum replied:
The `_MAX_ALLOWED` macros are all guarded by `PLATFORM` macros, so I don't think there is a problem. Historically, this is the pattern we've used for workarounds that are not necessary in perpetuity, and we have an existing example of this pattern in `PlatformHave.h` right now.

smfr commented on Source/WTF/wtf/PlatformUse.h:
```diff
@@ -376,3 +376,10 @@
+    || (PLATFORM(IOS_FAMILY) && __IPHONE_OS_VERSION_MAX_ALLOWED < 180000) \
+    || (PLATFORM(APPLETV) && __TV_OS_VERSION_MAX_ALLOWED < 180000) \
+    || (PLATFORM(WATCHOS) && __WATCH_OS_VERSION_MAX_ALLOWED < 110000)
+#define USE_CORE_TEXT_OPTICAL_SIZING_WORKAROUND 1
```
Workaround for what? A radar link in a comment would be good.

litherum committed:
[Cocoa] font-variation-settings:opsz is not applied when text-rendering:optimizeLegibility is set
https://bugs.webkit.org/show_bug.cgi?id=254897
rdar://106814619

Reviewed by Simon Fraser.

text-rendering:optimizeLegibility turns on kCTFontOpticalSizeAttribute:”auto” in Core Text,
and font-variation-settings:opsz turns on kCTFontVariationAttribute:opsz in Core Text. When
both are applied, font-variation-settings is supposed to win. However, in Core Text,
kCTFontOpticalSizeAttribute wins.

The solution for this is to set kCTFontOpticalSizeAttribute to a specific size, given by
font-variation-settings:opsz. This should only happen when OpticalSizingTypes::Everything
is enabled, so this patch adds an optional<float> to that struct. All this is guarded by
a USE() so we can remove it if/when Core text changes to match the CSS behavior.

Test: fast/text/variations/optical-sizing-optimizeLegibility.html

* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/platform/graphics/FontTaggedSettings.h:
* Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.cpp:
(WebCore::UnrealizedCoreTextFont::addAttributesForOpticalSizing):
(WebCore::UnrealizedCoreTextFont::modifyFromContext):
* Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.h:

Canonical link: https://commits.webkit.org/262831@main
